lib/mobile_application_platform/security_token/service.rb

```diff
@@ -16,9 +16,19 @@ def token(application:, icn:)
         Rails.logger.info("#{config.logging_prefix} token success, application: #{application}, icn: #{icn}")
         parse_response(response, application, icn)
       rescue Common::Client::Errors::ClientError => e
+        description = case e
+                      in _msg, 500, _body
+                        'service returned Internal Server error most likely due to an invalid ICN'
+                      in _msg, (400|401), { error: }
+                        error
+                      in _msg, (400|401), { error_description: }
```

Comment:
the pattern matching here is clever, but I'd argue we're making this unnecessarily brittle, conveying the errors we know (which we probably don't need much information about), and losing information about the errors we don't know (which we probably do need information about). Also if we get a status other than a `500/400/401`, the only guidance we'll get here is an `unknown` description, which probably is less useful than letting the original error itself be raised

Reply:
I would review the different ways this service appears to fail and just accommodate those, don't worry about `error_description` (especially since I'm pretty sure that was my mistake and will never be returned by this service).

Basically I'd just extract the status and plop it in the new error description, extract the description and plot it in the new error description, add additional context, and call it a day

Reply:
Yes I do understand if the API had a better error handling it would be better. I was just trying to giving some light to the errors the API may have.